### ISSUE TITLE
Log output gets buffered? 

### ISSUE THREAD

**Issue by RedBearAK** [joshgoebel/keyszer#58]
@joshgoebel 

Just wondering if you have any idea why the log output in the terminal (with `-v`) seems to get buffered somehow, with nothing showing, and then suddenly a ton of lines will appear later. There seems to be no real pattern to this. 

If I hit the debug key to kill the process, the final contents of the log will always appear immediately on screen. Showing events that often happened tens of seconds or longer in the past. 

I can usually hold down a modifier key for a while and force more output to appear, but I have to hold it for a few seconds, and it doesn't always show the entire latent output. I haven't found any particular type of event that always gets the log output to show everything available. 

The log has been doing this for me since the beginning. I just sort of ignored it because it would always appear eventually, or every time I hit the debug key. Hoping there's a way to make the output more regular, if not real-time.

**Comment by joshgoebel** [comment]
How are you running keyszer?

**Comment by joshgoebel** [comment]
I mean buffering is the default, but I actually haven't seen a problem running this in a terminal window so I wonder if you're running it otherwise?  You could try:

https://www.geeksforgeeks.org/python-sys-stdout-flush/

Adding flushing to the logger...

**Comment by joshgoebel** [comment]
Oh you said in the terminal. That's really weird never seen that

**Comment by RedBearAK** [comment]
This was happening in the `venv` running in a terminal tab, but now I'm on the v0.6.0 release and running it at startup with a script, but then killing it and running in the terminal again while testing/troubleshooting the new function(s). 

Tried putting a number of the flush commands in the config file both "bare" and inside the functions after `print()` commands. There's no error when I load the config, but it also doesn't seem to make any discernible difference in the lengthy delay before output appears. 

I am doing `tail -f` and filtering a lot of lines other than my printed output and lines starting with `(DD)` for instance. But I know a lot of the lines that I can "force" to appear by holding down a mod for several seconds (often have to do this twice or just repeatedly tap a mod) happened quite a while earlier so I would normally expect them to just immediately appear in my `tail` as soon as they are appended to the output file. And I think they do, so it's more like they aren't even in the output file for anywhere from seconds to minutes, until I do some kind of event that seems to trigger a real flush of the output from wherever it's hiding. 

I thought maybe the `2>&1` redirect might have something to do with it, but I've tried taking that out and it didn't seem to make any difference either. 

This is the command I typically run. It creates a dated output file and a "latest" output so that I can just keep using the same command in another tab to `tail` the newest output. 

```
keyszer -v -c ~/.config/kinto/keyszer.py 2>&1 | tee ~/Documents/Dev-Projects/keyszer-debug/keyszer-output-latest.txt | tee ~/Documents/Dev-Projects/keyszer-debug/keyszer-output-`date +%F-%H%M%S`.txt
```

And to monitor in another tab, compacting by removing blank lines and `(II)`/`(OO)` and repeats: 

```
tail -f -n 1000 ~/Documents/Dev-Projects/keyszer-debug/keyszer-output-latest.txt | grep -v "^$" | grep -v "^(II" | grep -v "^(OO" | grep -v "repeat"
```

Yes, it's tailing the last 1,000 lines of the config, to make sure I catch the beginning of a restart cycle in case there's a ton of events happening right away. As far as I know from previous experience with `tail` it should be able to load thousands of lines in the blink of an eye, so that shouldn't be the source of any kind of delay. 

Also, like I said, this happens in the terminal tab where `keyszer` is running and directly displaying the raw output without any tailing or `grep` filtering. It just seems to completely stop at random times and places, including sometimes at the very beginning, not showing individual keyboard events, then suddenly blasting hundreds of new lines into the terminal. 

Maybe there's something funky about the way the buffer works in GNOME Terminal. I'll try a different terminal. 

Or there's something weird in my config file and it's been there since the beginning. The fact that it works fine otherwise is not necessarily proof that it isn't the source of the behavior. But I completely disabled the custom `sleep_and_then` and `launch_and_then` functions from Lenbok that I had been playing with in the past, so I don't know of anything else I've ever added to the config that could possibly affect the log output. All other customizations should just be normal `K()` shortcut lines. 

> That's really weird never seen that

Thanks, I do my best. 🤘 😜

**Comment by joshgoebel** [comment]
> I am doing tail -f and filtering a lot of lines 

The `tail` will break things, but adding flushes everywhere should fix it... (I think)

> this happens in the terminal tab where keyszer is running and directly displaying the raw output without any tailing or grep filtering. 

Ok, I have zero explanation for that... when I see these issues is usually with `pipe` where the buffers between the two (or more) pipes causes issues... I have as far as I can tell absolute zero lag when just running raw in a terminal (alacritty).

**Comment by joshgoebel** [comment]
> This is the command I typically run. `tee...tee...`

If that's what you mean by "directly" that's pretty indirect... using `tee` twice means TWO buffers that can get behind... one of them might kill the flush from the first, really not sure.

**Comment by RedBearAK** [comment]
As far as I know `tee` just copies what's going through the pipe, while the usual output just keeps on going past it. It shouldn't be capable of introducing any kind of delay. I've never seen it delaying anything in other usage. But I am no terminal expert. 

I'm having issues with Snap apps, and that's the only method I have to install alacritty in the usual ways, without grabbing it directly from GitHub or something. But I tried `kitty` which I believe is closely related, and see the same thing happening. 

I'm leaning more toward my config file doing something odd. Don't know what else it could possibly be. 

Not going to worry too much about it for now.

**Comment by RedBearAK** [comment]
@joshgoebel 

PR #78 should clear up this "buffering" issue for the most part, as long as the built-in logging functions are used. Terminal logging output is very responsive for me after adding flushing to all the print commands in the logger. I was doing this with just random print commands I was using for peeking into things, but doing centrally it in the logger seems like a better idea.

**Comment by RedBearAK** [comment]
Closing as resolved by PR #79.
